_posts/2025-07-16-oracle-does-not-contain-scn.adoc

```diff
@@ -0,0 +1,169 @@
+
+* Long-running queries or transactions
+* Point-in-time recovery
+* Downstream systems like Debezium that need historical log data
+
+=== Redo Threads
+
+Whether you've deployed Oracle as a Standalone instance or if you're using Oracle RAC, each Oracle instance writes to its own redo thread.
+While SCNs increment globally, each redo thread maintains a logical series of chronological changes that have occurred on that specific instance.
+In practice, this means:
+
+* One redo thread (T1) may contain changes from transactions that logically happen _after_ those in another redo thread (T2), even if the SCNs seem interleaved.
+* Debezium and by extension LogMiner, needs access to logs across all redo threads to correctly reconstruct the global order of transactions in the database.
+* Missing or unarchived redo logs for _any redo thread_ can cause gaps, which Debezium guards against by performing log consistency check when Oracle performs a log switch (which we'll discuss more below)
+
+== How to troubleshoot the "SCN not found" error?
+
+There are several points during the Debezium connector lifecycle where the Oracle connector may report that it cannot find a specific SCN in the logs.
+Let's break down where this happens and how to troubleshoot each scenario.
+
+=== Log position validation
+
+When a Debezium Oracle connector starts, one of the first steps is validating that the resume position, based on the last known SCN in the offset, is still valid.
+To do this, Debezium compares the stored offset SCN against the earliest available SCN in the database’s redo logs (online or archived).
+
+The logic is straightforward:
+
+* If the earliest SCN ≤ offset SCN, the position is valid.
+* If the earliest SCN > offset SCN, the position is invalid, and Debezium throws an error stating the SCN cannot be found.
+
+==== Why would the SCN become invalid?
+
+There are several common causes:
+
+Short archive log retention::
+The Oracle archive policy may be purging logs faster than Debezium can process them.
+Extended connector downtime::
+If the connector is offline longer than the log retention period, the required logs may no longer exist.
+Skewed activity between captured and non-captured tables::
+The global SCN increases as transactions occur, even if Debezium isn’t capturing those tables. This can result in a stale SCN in the offsets.
+
+==== What you can do
+
+The following is what you can do depending on the use case.
+
+Archive logs purged too early::
+Work with DBAs to increase the retention window or configure a secondary archive destination with longer retention.
+Ensure Debezium has access to both online and archive logs from all redo threads.
+
+Connector is offline for too long::
+Ask DBAs to restore the missing archive logs if possible. Once restored, Debezium should resume without changes.
+If restoration isn't possible: +
+  * Manually update the connector offsets to a valid SCN that exists in the logs.
```

Comment:
Maybe we could point to https://github.com/debezium/debezium-examples/tree/main/offset-editor?